atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1149,26 +1186,68 @@ private int estimateSize(List<RowResult<Value>> rangeRows) {
 
         getMeter(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_CELLS_READ, tableRef).mark(rawResults.size());
 
+        // LinkedList is chosen for fast append operation since we just add to this collection.
+        Collection<Map.Entry<Cell, T>> resultsAccumulator = new LinkedList<>();
+
         if (AtlasDbConstants.HIDDEN_TABLES.contains(tableRef)) {
             Preconditions.checkState(allowHiddenTableAccess, "hidden tables cannot be read in this transaction");
             // hidden tables are used outside of the transaction protocol, and in general have invalid timestamps,
             // so do not apply post-filtering as post-filtering would rollback (actually delete) the data incorrectly
             // this case is hit when reading a hidden table from console
             for (Map.Entry<Cell, Value> e : rawResults.entrySet()) {
-                resultsAccumulator.put(e.getKey(), transformer.apply(e.getValue()));
+                resultsAccumulator.add(
+                        Maps.immutableEntry(e.getKey(), transformer.apply(e.getValue())));
             }
-            return resultsAccumulator;
+            return Futures.immediateFuture(resultsAccumulator);
         }
 
-        Map<Cell, Value> remainingResultsToPostfilter = rawResults;
         AtomicInteger resultCount = new AtomicInteger();
-        while (!remainingResultsToPostfilter.isEmpty()) {
-            remainingResultsToPostfilter = getWithPostFilteringInternal(
-                    tableRef, remainingResultsToPostfilter, resultsAccumulator, resultCount, transformer);
-        }
 
-        getMeter(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_CELLS_RETURNED, tableRef).mark(resultCount.get());
-        return resultsAccumulator;
+        return Futures.transformAsync(
+                Futures.immediateFuture(rawResults),
+                remainingResultsToPostFilter ->
+                        getWithPostFilteringIterate(tableRef,
+                                remainingResultsToPostFilter,
+                                resultCount,
+                                resultsAccumulator,
+                                transformer,
+                                asyncKeyValueService,
+                                asyncTransactionService),
+                MoreExecutors.directExecutor());
+    }
+
+    private <T> ListenableFuture<Collection<Map.Entry<Cell, T>>> getWithPostFilteringIterate(
+            TableReference tableReference,
+            Map<Cell, Value> remainingResultsToPostFilter,
+            AtomicInteger resultCounter,
+            Collection<Map.Entry<Cell, T>> resultsAccumulator,
+            Function<Value, T> transformer,
+            AsyncKeyValueService asyncKeyValueService,
+            AsyncTransactionService asyncTransactionService) {
+        if (remainingResultsToPostFilter.isEmpty()) {
+            getMeter(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_CELLS_RETURNED, tableReference).mark(resultCounter.get());
```

Comment:
isn't this just the same as the size of the `resultsAccumulator`? If so we can get rid of the `resultCounter` atomic integer!